:seedling: refactor SortForCreate to use sort.Slice

kokes opened the pull request:
**What this PR does / why we need it**:

I found [SortForCreate](https://github.com/kubernetes-sigs/cluster-api/blob/4abf44cd85c4590602e4c10543d53cd4ec914845/util/resource/resource.go#L50) to be very inefficient, it's basically O(n^2), it also needlessly allocates. In order to keep the contract (even though this exported function [is not widely used](https://github.com/search?type=code&q=SortForCreate+language%3AGo&l=Go&p=1)), I didn't change this function's behaviour (i.e. its purity), ~~but I did deprecate it~~. (It could've been even faster if we opted for an in-place sort, but I didn't want to change the function's behaviour.)

The function doesn't appear to be used in any tight loops or with a large number of elements, so it's a low prio PR, but it's something that could bite people applying a large number of small resources (see how it falls off a cliff with 10k elements).

**Benchmark comparison**
```
name             old time/op    new time/op    delta
Sort/sort-10       10.0µs ±31%     3.7µs ±55%  -63.11%  (p=0.000 n=10+10)
Sort/sort-100       423µs ±10%      58µs ±28%  -86.35%  (p=0.000 n=10+10)
Sort/sort-1000     45.0ms ±13%     0.5ms ±12%  -98.83%  (p=0.000 n=10+9)
Sort/sort-10000     5.90s ±22%     0.01s ±30%  -99.86%  (p=0.000 n=10+10)

name             old alloc/op   new alloc/op   delta
Sort/sort-10         248B ± 0%      136B ± 0%  -45.16%  (p=0.000 n=10+10)
Sort/sort-100      2.04kB ± 0%    0.95kB ± 0%  -53.33%  (p=0.000 n=10+10)
Sort/sort-1000     25.2kB ± 0%     8.2kB ± 0%  -67.28%  (p=0.000 n=10+10)
Sort/sort-10000     358kB ± 0%      82kB ± 0%  -77.08%  (p=0.000 n=10+10)

name             old allocs/op  new allocs/op  delta
Sort/sort-10         5.00 ± 0%      3.00 ± 0%  -40.00%  (p=0.000 n=10+10)
Sort/sort-100        8.00 ± 0%      3.00 ± 0%  -62.50%  (p=0.000 n=10+10)
Sort/sort-1000       12.0 ± 0%       3.0 ± 0%  -75.00%  (p=0.000 n=10+10)
Sort/sort-10000      20.0 ± 0%       3.0 ± 0%  -85.00%  (p=0.000 n=10+10)
```

**GH issue**
Fixes #9253

/area util

killianmuldoon reviewed (commented):
/ok-to-test

killianmuldoon commented on util/resource/resource.go:
```diff
@@ -17,61 +17,56 @@ limitations under the License.
+// Deprecated: use PriorityLess with sort.Slice, sort.SliceStable
+// instead for fast in-place sorting.
 func SortForCreate(resources []unstructured.Unstructured) []unstructured.Unstructured {
-	var ret []unstructured.Unstructured
```
If we're deprecating this function we should keep the behaviour the same.

killianmuldoon replied:
And we should update the current usages in CAPI itself to use the new version.

kokes replied:
> If we're deprecating this function we should keep the behaviour the same.

In what way is it different? Are you referring to the `resources == nil` case? In that case I return a 0-sized slice whereas it currently returns nil, correct?

killianmuldoon replied:
Sorry - I was unclear there. 

I don't think this function should change if this PR deprecates it.

Alternatively it could be retained and the implementation be changed to using PriorityLess with sort.Slice. 

We should choose between deprecating the function or changing the implementation. Doing both in this PR doesn't make sense to me.

kokes replied:
Oh I see - deprecation shouldn't result in a refactor.

kokes replied:
Removed the deprecation marker, because its API is much easier to work with than my `sort.Slice` alternative.

kokes commented:
I redid it a bit - I decided not to deprecate `SortForOrder` after all, because the `sort.Slice` usage is quite clunky (due to the closure that's required) and should only be used in a performance-critical scenario (and anyone can use it, should they need it). The new (allocating) implementation is good enough, so I'll leave it that way.

We could also:

1. Unexport `PriorityLess` to keep the API surface smaller.
2. Create a function to satisfy [SortFunc](https://pkg.go.dev/slices#SortFunc), though that API so super new that it won't be leveraged until 1.21 is the oldest Go version supported, which is probably quite far away atm.

Neither is a priority tho.

killianmuldoon commented:
/ok-to-test

enxebre commented on util/resource/resource.go:
```diff
@@ -17,61 +17,54 @@ limitations under the License.
+
+// PriorityLess returns true if o1 should be created before o2.
+// To be used in sort.{Slice,SliceStable} to sort manifests in place.
+func PriorityLess(o1, o2 unstructured.Unstructured) bool {
```
nit: any reason for this to be exported?

kokes replied:
It was originally exported, so that we could use it with `sort.Slice` across the project... but since we're not doing that, I actually [suggested](https://github.com/kubernetes-sigs/cluster-api/pull/9251#issuecomment-1686151711) we can unexport it to make the API cleaner.

kokes replied:
Unexported.

richardcase reviewed (commented):
The behaviour of the `SortForCreate` stays the same. Interestingly, the original implementation doesn't match the comments and the tests don't cover the scenarios.

For example, the comments say that `Pods go before Replicaset` but if you extend the test to include pods, deployments, replicasets like this:

```go
	cm := unstructured.Unstructured{}
	cm.SetKind("ConfigMap")

	ns := unstructured.Unstructured{}
	ns.SetKind("Namespace")

	ep := unstructured.Unstructured{}
	ep.SetKind("Endpoint")

	dp := unstructured.Unstructured{}
	dp.SetKind("Deployment")

	rs := unstructured.Unstructured{}
	rs.SetKind("ReplicaSet")

	pd := unstructured.Unstructured{}
	pd.SetKind("Pod")

	resources := []unstructured.Unstructured{ep, cm, ns, dp, rs, pd}
```

The replicaset is sorted before the pod. The issue is mainly that in the original implementation the slice should have contained `Pod` & `Endpoint` instead of `Pods` & `Endpoints`.

richardcase commented:
@killianmuldoon @enxebre - do we care that the original behaviour doesn't match the comments in the code? I see we use it in the CRS and clusterctl (for cert-manager) code but i don't have any context on its original inclusion. 

(If we do care i'd say its out of scope for this PR though)

killianmuldoon commented:
>@killianmuldoon @enxebre - do we care that the original behaviour doesn't match the comments in the code? I see we use it in the CRS and clusterctl (for cert-manager) code but i don't have any context on its original inclusion. 

IMO it's clear the current behaviour is a bug which could actually cause issues in e.g. a ClusterResourceSet. Fine with properly fixing that here so long as we update the tests to cover the full behavior of the sort.

killianmuldoon reviewed (commented):
Looks good - just a couple of asks for the tests.

Could you update the test to add all of the types listed in resources, and a couple that aren't e.g. `Deployment'. It would also be good to define them in a map and convert that to an input sice to ensure we get multiple permutations in the test.

killianmuldoon commented on util/resource/resource_test.go:
```diff
@@ -41,3 +62,21 @@ func TestSortForCreate(t *testing.T) {
 	g.Expect(sorted[1].GetKind()).To(BeIdenticalTo("ConfigMap"))
 }
+
+func BenchmarkSort(b *testing.B) {
```
Let's remove the benchmarking tests as it doesn't actually `test` for anything without being run manually.

kokes commented:
I've pushed two commits, where I:

- unexported `PriorityLess` since we don't use it anywhere othere than this package
- removed the benchmark since we don't run benchmarks in this projects apparently
- fixed the plural of `Pods` and `Endpoints`
- reworked the existing test to include elements with no priority to check they are assigned least priority
- added a new, more comprehensive test, where I add all sorts of resources, repeatedly shuffle and check that sorting these always results in all the kinds with priorities frontloaded and the rest of resources with no particular order

killianmuldoon reviewed (commented):
Looks great! Thanks for the testing improvements. 

Could you squash the commits?

kokes commented:
Done

killianmuldoon commented:
Thanks again!

/approve

sbueringer commented on util/resource/resource.go:
```diff
@@ -17,61 +17,54 @@ limitations under the License.
+var priorityMap = map[string]int{
 	// Namespaces go first because all namespaced resources depend on them.
-	"Namespace",
+	"Namespace": -100,
```
It seems a bit strange that the kind with the highest priority has the lowest priority value

sbueringer replied:
Apart from that, nice improvement!

sbueringer replied:
@kokes Can we have a quick follow-up PR to align priority values to priority?

kokes replied:
This is so that values are sorted naturally without having to fiddle with anything. I thought about giving these positive values, which would then require the `Less` function to have the inequality check reverted... I thought it'd be a bit confusing. But if it sounds confusing to have it this way, fine, we can flip it, the tests should cover this nicely.

sbueringer replied:
I just thought if the map is the "priority map" then I would expect the kind with the highest priority to come first not last

sbueringer replied:
If it would be called the orderMap it's perfectly fine I guess :)

kokes replied:
Like this? https://github.com/kubernetes-sigs/cluster-api/compare/main...kokes:sort_order_naming?expand=1

I kept `priorityLess`, even though that might sound confusing - but it's for sorting by priority and it's the `Less` function for native Go sorting, so I thought I'd get away with that.

sbueringer replied:
Yeah let's go with that
